### PR TITLE
ports - sav

### PR DESCRIPTION
**Pull request by qqdipps**
# JS Adagrams
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
--- | ---
What patterns were you able to use from your Ruby knowledge to apply to JavaScript? | Enumerables: sort, find, map. Loops (syntax a little different), logical boolean statements, if/else if /else blocks, recursion.  
Did you need to use different strategies to find information online about JavaScript than you would for Ruby? | Different sources; had to look lots of stuff up to see if it was possible and what the 'JavaScript way' was. 
What was a challenge you were able to overcome on this assignment? | I created a nested helper object inside Adagrams and found out that there is no way to reference parent object `this`, but it worked out well since I found that I could use `parentObjectName.parentMethod();` to reference the outer method from the nested object.
What has been interesting and positive about learning a new programming language? | It was super exciting and shockingly fast. 
What is something to focus on your learnings in JavaScript in the next week? | practice, experimentation, and keep reading the documentation!

### CODE REVIEW

**Comment by tildeee** [comment]
# JS Adagrams
## What We're Looking For



Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | x
Small commits with meaningful commit messages | x
**Code Requirements** |
`drawLetters` method | x
Uses appropriate data structure to store the letter distribution | x
All tests for `drawLetters` pass | x
`usesAvailableLetters` method | x
All tests for `usesAvailableLetters` pass | x
`scoreWord` method | x
Uses appropriate data structure to store the letter scores | x
All tests for `scoreWord` pass | x
`highestScoreFrom` method | x
Appropriately handles edge cases for tie-breaking logic | x
All tests for `highestScoreFrom` pass | x
**Overall** |



Hi Sav! Great work on this project!


As I think you may know, you definitely went the route of going for an interesting solution rather than a readable one. I'm really happy about it! In general though, past Ada, my assumption will be that unless you have a good reason to have extremely optimal code, I'd tend towards readability and leaning on existing built-in functionality (like using `includes`).


That being said, I love this code! It's really fun and interesting and exciting.


I've left a looot of comments! All of my comments are honestly along the themes of code-style, readability, really being nitpicky about line-to-line, and how to make it more JavaScript-y. I wouldn't say that any of the comments or feedback are causing me any concern.


Well done! Let me know if you have any questions.

File: src/adagrams.js [hunk 1]
```diff
@@ -1,6 +1,134 @@
...
+                  s: 1, t: 1, d: 2, g: 2, l: 1, b: 3, c: 3, 
+                  m: 3, p: 3, f: 4, h: 4, v: 4, w: 4, y: 4, 
+                  k: 5, j: 8, x: 8, q: 10, z: 10},
+
   drawLetters() {
-    // Implement this method for wave 1
+    let drawnLetters = [];
+    for (let i = 0; i < 10; i += 1) {
+      const letterCount = this.letters.length - i;
+      const randomIndex = Math.floor(Math.random() * Math.floor(letterCount));
```
Comment (tildeee) [inline src/adagrams.js:24]: I'm fairly certain that you don't need the inner `Math.floor`

File: src/adagrams.js [hunk 2]
```diff
@@ -1,6 +1,134 @@
...
+      const selectIndex = letterCount - 1 - i;
+      const temp = this.letters[selectIndex];
+      this.letters[selectIndex] = this.letters[randomIndex];
+      this.letters[randomIndex] = temp;
```
Comment (tildeee) [inline src/adagrams.js:28]: In the Adagrams project description, we specify to not modify the letter pool in a way that would affect future calls to `drawLetters` ;--)

File: src/adagrams.js [hunk 3]
```diff
@@ -1,6 +1,134 @@
...
+      drawnLetters.push(this.letters[selectIndex]);
+    }
+    return drawnLetters;
+  },
+
+  usesAvailableLetters(input, lettersInHand) {
+    input = input.split('').sort();
+    lettersInHand = lettersInHand.sort();
+    for (let i = 0, j = 0; i < input.length; ) {
+      if (!lettersInHand[j]) return false;
```
Comment (tildeee) [inline src/adagrams.js:38]: Hehehe, yes, JS supports `if` statements with no curly-braces if they are one-line! I am a little nervous about the habit of this, because I am a protective coder, so I'll just mention that this is risky ;-)

File: src/adagrams.js [hunk 4]
```diff
@@ -1,6 +1,134 @@
...
+        j += 1;
+      } else {
+        return false;
+      }
+    }
+    return true;
+  },
+
+  scoreWord(word) {
+    if (word.length === 0 || !word) return 0;
```
Comment (tildeee) [inline src/adagrams.js:52]: If you are really checking for the case if `word` is `null`, you should check for that case first before you call `.length` on it, so: `( !word || word.length === 0 )`

File: src/adagrams.js [hunk 5]
```diff
@@ -1,6 +1,134 @@
...
+    //prettier-ignore
+    const score = word.split('').map(char => this.letterPoints[char.toLowerCase()]);
```
Comment (tildeee) [inline src/adagrams.js:54]: HOORAY! `map`!! :D

Comment (tildeee) [inline src/adagrams.js:54]: Since this represents an array, it probably reads better to name this `scores`

File: src/adagrams.js [hunk 6]
```diff
@@ -1,6 +1,134 @@
...
+    if (score.length >= 7) {
+      score[0] += 8;
+    }
+    return score.reduce((a, b) => a + b);
```
Comment (tildeee) [inline src/adagrams.js:58]: I LOVE A REDUCE!

File: src/adagrams.js [hunk 7]
```diff
@@ -1,6 +1,134 @@
...
+      this.helper.sortWordsByScore(wordScorePairs)
+    );
+    // No tie, return highest scoring word
+    if (wordTies.length === 1) return wordTies[0];
+    // determine eligible winning word
+    const eligibileWinningWord = this.helper.getEligibileWinningWord(
+      this.helper.sortWordsByLength(wordTies)
+    );
+    // ensures first occurence of an eligibile winning word from orginal list of words is returned.
+    return this.helper.findFirstWinningWord(
```
Comment (tildeee) [inline src/adagrams.js:76]: Hm... I am not convinced that we can't just `return eligibleWinningWord` at this point. If you'd like to prove me wrong, then I'd love to see the unit test that would break without the call to `this.helper.findFirstWinningWord` ;-)

File: src/adagrams.js [hunk 8]
```diff
@@ -1,6 +1,134 @@
...
+      eligibileWinningWord.word.length,
+      eligibileWinningWord.score,
+      wordScorePairs
+    );
+  },
+
+  helper: {
```
Comment (tildeee) [inline src/adagrams.js:83]: I like this! It's cool organization. However, I feel like it's worth mentioning-- I don't particularly know if having a member in an object named `helper` is aligned with any sort of design pattern I've seen in the wild. It may make sense to just make these regular ole methods on `Adagrams` or something similar!

File: src/adagrams.js [hunk 9]
```diff
@@ -1,6 +1,134 @@
...
+    mapScores(words) {
+      return words.map(word1 => {
```
Comment (tildeee) [inline src/adagrams.js:85]: Why not name this `word` instead of `word1`?

File: src/adagrams.js [hunk 10]
```diff
@@ -1,6 +1,134 @@
...
+        let wordPair = {};
```
Comment (tildeee) [inline src/adagrams.js:86]: weirdly enough, this can be a `const`. Regardless of variable keyword, object members can be modified. Not that I'm happy about it, but I thought you'd appreciate knowing!

File: src/adagrams.js [hunk 11]
```diff
@@ -1,6 +1,134 @@
...
+        wordPair.word = word1;
+        wordPair.score = Adagrams.scoreWord(word1);
+        return wordPair;
```
Comment (tildeee) [inline src/adagrams.js:89]: I think that this function would be a lot cleaner if we went ahead and returned the object literal, so: `return { word: word1, score: Adagrams.scoreWord(word1) };`

File: src/adagrams.js [hunk 12]
```diff
@@ -1,6 +1,134 @@
...
+    },
+
+    getTies(scores, index = 0, keeper = null) {
+      keeper = keeper || [scores[index]];
+      const scoreChecker = scores[index + 1] ? scores[index + 1].score : null;
+      if (!keeper || scores[index].score !== scoreChecker) {
+        return keeper;
+      } else {
+        keeper.push(scores[index + 1]);
+        return this.getTies(scores, index + 1, keeper);
```
Comment (tildeee) [inline src/adagrams.js:104]: FANCY!

File: src/adagrams.js [hunk 13]
```diff
@@ -1,6 +1,134 @@
...
+      if (sortedWordPairTies[0].word.length === 10) {
+        wordScorePair = sortedWordPairTies[0];
+      } else {
+        wordScorePair = sortedWordPairTies[sortedWordPairTies.length - 1];
+      }
+      return wordScorePair;
+    },
+
+    findFirstWinningWord(wordLength, score, wordScorePairs) {
+      const winningWord = wordScorePairs.find(
```
Comment (tildeee) [inline src/adagrams.js:125]: Nice use of `find`!

File: src/adagrams.js (hunk 2 above)
Comment (qqdipps) [reply src/adagrams.js:28]: The intention was to swap letters that were drawn to end of the array then temporarily "shrink" available letters `const letterCount = this.letters.length - i;` after each iteration by the number of iterations that is the number of letters drawn. Since it uses a random index, does rearranging array affect future calls?

File: src/adagrams.js (hunk 4 above)
Comment (qqdipps) [reply src/adagrams.js:52]: Yes most def!

File: src/adagrams.js (hunk 1 above)
Comment (qqdipps) [reply src/adagrams.js:24]: yuppers!

File: src/adagrams.js (hunk 7 above)
Comment (qqdipps) [reply src/adagrams.js:76]: I think you're right. I had thought that since I was sorting by score and then later by length it would mess up the order of the words. In retrospect, I see that though the order may change, the relative order of two or more winning words would stay the same?

File: src/adagrams.js (hunk 8 above)
Comment (qqdipps) [reply src/adagrams.js:83]: I realized it was odd while I was doing it =P. It felt like I had a bunch of methods that were about helping find the winning word that didn't feel like they should be able to be accessed directly by an adagrams object, so I thought that by putting the methods in a helper object they felt less accessible and more private. If this was a class and JS allowed private methods that's what I would have done. =)

File: src/adagrams.js (hunk 9 above)
Comment (qqdipps) [reply src/adagrams.js:85]: I was concerned about it since I was using word as a key ` wordPair.word = word1;`I prob should have just explored and see if it would have been okay to do.

File: src/adagrams.js (hunk 11 above)
Comment (qqdipps) [reply src/adagrams.js:89]: Yes! I had trouble with this. I think that I forgot to explicitly return and I may have been missing a set of curly braces, so it morphed into a broken down way of going about it.  Def a workaround, but now I see the light. ;-)
